Update code to use functions which are not deprecated in OpenSSL 3

--- azure-uamqp-python.orig/src/vendor/azure-uamqp-c/deps/azure-c-shared-utility/adapters/tlsio_openssl.c
+++ azure-uamqp-python/src/vendor/azure-uamqp-c/deps/azure-c-shared-utility/adapters/tlsio_openssl.c
@@ -451,6 +451,7 @@ static const IO_INTERFACE_DESCRIPTION tl
     tlsio_openssl_setoption
 };
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
 static LOCK_HANDLE * openssl_locks = NULL;
 
 
@@ -477,6 +478,7 @@ static void openssl_lock_unlock_helper(L
         }
     }
 }
+#endif
 
 static void log_ERR_get_error(const char* message)
 {
@@ -499,6 +501,7 @@ static void log_ERR_get_error(const char
     }
 }
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
 static STATIC_VAR_UNUSED struct CRYPTO_dynlock_value* openssl_dynamic_locks_create_cb(const char* file, int line)
 {
 #ifdef NO_LOGGING
@@ -542,10 +545,11 @@ static STATIC_VAR_UNUSED void openssl_dy
     Lock_Deinit(dynlock_value->lock);
     free(dynlock_value);
 }
+#endif
 
 static void openssl_dynamic_locks_uninstall(void)
 {
-#if (OPENSSL_VERSION_NUMBER >= 0x00906000)
+#if (OPENSSL_VERSION_NUMBER >= 0x00906000) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
     CRYPTO_set_dynlock_create_callback(NULL);
     CRYPTO_set_dynlock_lock_callback(NULL);
     CRYPTO_set_dynlock_destroy_callback(NULL);
@@ -554,13 +558,14 @@ static void openssl_dynamic_locks_uninst
 
 static void openssl_dynamic_locks_install(void)
 {
-#if (OPENSSL_VERSION_NUMBER >= 0x00906000)
+#if (OPENSSL_VERSION_NUMBER >= 0x00906000) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
     CRYPTO_set_dynlock_destroy_callback(openssl_dynamic_locks_destroy_cb);
     CRYPTO_set_dynlock_lock_callback(openssl_dynamic_locks_lock_unlock_cb);
     CRYPTO_set_dynlock_create_callback(openssl_dynamic_locks_create_cb);
 #endif
 }
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
 static void STATIC_VAR_UNUSED openssl_static_locks_lock_unlock_cb(int lock_mode, int lock_index, const char * file, int line)
 {
     if (lock_index < 0 || lock_index >= CRYPTO_num_locks())
@@ -572,9 +577,11 @@ static void STATIC_VAR_UNUSED openssl_st
         openssl_lock_unlock_helper(openssl_locks[lock_index], lock_mode, file, line);
     }
 }
+#endif
 
 static void openssl_static_locks_uninstall(void)
 {
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     if (openssl_locks != NULL)
     {
         int i;
@@ -595,12 +602,14 @@ static void openssl_static_locks_uninsta
     {
         LogError("Locks already uninstalled");
     }
+#endif
 }
 
 static int openssl_static_locks_install(void)
 {
     int result;
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     if (openssl_locks != NULL)
     {
         LogInfo("Locks already initialized");
@@ -645,6 +654,9 @@ static int openssl_static_locks_install(
             }
         }
     }
+#else
+    result = 0;
+#endif
     return result;
 }
 
@@ -774,7 +786,11 @@ int engine_load(TLS_IO_INSTANCE* tls)
     int result;
 
     // TLSIO is only responsible for the Engine structural reference.
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
     tls->engine = ENGINE_by_id(tls->engine_id);
+#else
+    tls->engine = NULL;
+#endif
     if (tls->engine == NULL)
     {
         log_ERR_get_error("unable to load engine by ID.");
@@ -1234,11 +1250,17 @@ static int create_openssl_instance(TLS_I
 
 int tlsio_openssl_init(void)
 {
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     (void)SSL_library_init();
 
     SSL_load_error_strings();
+#endif
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
     ERR_load_BIO_strings();
+#endif
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     OpenSSL_add_all_algorithms();
+#endif
 
     if (openssl_static_locks_install() != 0)
     {
@@ -1257,10 +1279,16 @@ void tlsio_openssl_deinit(void)
 #if  (OPENSSL_VERSION_NUMBER >= 0x00907000L) && (FIPS_mode_set)
     FIPS_mode_set(0);
 #endif
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     CRYPTO_set_locking_callback(NULL);
+#endif
+#if (OPENSSL_VERSION_NUMBER < 0x10000000L)
     CRYPTO_set_id_callback(NULL);
+#endif
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     ERR_free_strings();
     EVP_cleanup();
+#endif
 
 #if   (OPENSSL_VERSION_NUMBER < 0x10000000L)
     ERR_remove_state(0);
@@ -1270,7 +1298,9 @@ void tlsio_openssl_deinit(void)
 #if  (OPENSSL_VERSION_NUMBER >= 0x10002000L) &&  (OPENSSL_VERSION_NUMBER < 0x10010000L) && (SSL_COMP_free_compression_methods)
     SSL_COMP_free_compression_methods();
 #endif
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     CRYPTO_cleanup_all_ex_data();
+#endif
 }
 
 CONCRETE_IO_HANDLE tlsio_openssl_create(void* io_create_parameters)
@@ -1743,7 +1773,9 @@ int tlsio_openssl_setoption(CONCRETE_IO_
         #ifndef OPENSSL_NO_ENGINE
         else if (strcmp(OPTION_OPENSSL_ENGINE, optionName) == 0)
         {
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
             ENGINE_load_builtin_engines();
+#endif
 
             if (mallocAndStrcpy_s((char**)&tls_io_instance->engine_id, value) != 0)
             {
--- azure-uamqp-python.orig/src/vendor/azure-uamqp-c/deps/azure-c-shared-utility/adapters/x509_openssl.c
+++ azure-uamqp-python/src/vendor/azure-uamqp-c/deps/azure-c-shared-utility/adapters/x509_openssl.c
@@ -138,6 +138,17 @@ static int load_ecc_key(SSL_CTX* ssl_ctx
 static int load_key_RSA(SSL_CTX* ssl_ctx, EVP_PKEY* evp_key)
 {
     int result;
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+    if (SSL_CTX_use_PrivateKey(ssl_ctx, evp_key) != 1)
+    {
+        log_ERR_get_error("Failure calling SSL_CTX_use_PrivateKey");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        result = 0;
+    }
+#else
     /*Codes_SRS_X509_OPENSSL_02_005: [ x509_openssl_add_credentials shall use PEM_read_bio_RSAPrivateKey to read the x509 private key. ]*/
     RSA* privatekey = EVP_PKEY_get1_RSA(evp_key);
     if (privatekey == NULL)
@@ -163,6 +174,7 @@ static int load_key_RSA(SSL_CTX* ssl_ctx
         }
         RSA_free(privatekey);
     }
+#endif
     return result;
 }
 
@@ -314,7 +326,12 @@ int x509_openssl_add_credentials(
         #ifndef OPENSSL_NO_ENGINE
         else if (x509privatekeytype == KEY_TYPE_ENGINE)
         {
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
             result = x509_openssl_add_engine_key(ssl_ctx, x509privatekey, engine);
+#else
+	    LogError("OpenSSL Engine not supported.");
+	    result = MU_FAILURE;
+#endif
         }
         #endif // OPENSSL_NO_ENGINE
         else
